syntax = "proto3";

package promo;

service PromoService {
  rpc CreatePromo(CreatePromoRequest) returns (PromoResponse) {}
  rpc GetPromo(GetPromoRequest) returns (PromoResponse) {}
  rpc UpdatePromo(UpdatePromoRequest) returns (PromoResponse) {}
  rpc DeletePromo(DeletePromoRequest) returns (DeleteResponse) {}
  rpc ListPromos(ListPromosRequest) returns (ListPromosResponse) {}
}

message CreatePromoRequest {
  string name = 1;
  string description = 2;
  string creator_id = 3;
  double discount = 4;
  string code = 5;
}

message PromoResponse {
  string id = 1;
  string name = 2;
  string description = 3;
  string creator_id = 4;
  double discount = 5;
  string code = 6;
  string created_at = 7;
  string updated_at = 8;
}

message GetPromoRequest {
  string id = 1;
}

message UpdatePromoRequest {
  string id = 1;
  string creator_id = 2;
  optional string name = 3;
  optional string description = 4;
  optional double discount = 5;
  optional string code = 6;
}

message DeletePromoRequest {
  string id = 1;
  string creator_id = 2;
}

message ListPromosRequest {
  string creator_id = 1;
  int32 page = 2;
  int32 per_page = 3;
}

message ListPromosResponse {
  repeated PromoResponse promos = 1;
  int32 total = 2;
}

message DeleteResponse {
  bool success = 1;
}
